toolchain/common/data_index.h

```diff
@@ -0,0 +1,156 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_COMMON_DATA_INDEX_H_
+#define CARBON_TOOLCHAIN_COMMON_DATA_INDEX_H_
+
+#include "common/ostream.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/iterator.h"
+#include "llvm/Support/Format.h"
+
+namespace Carbon {
+
+template <typename DataType>
+class DataIterator;
+
+// A lightweight handle to an item in a SmallVector. These are constructed using
+// `Append(data, el)` then used with `index.In(data)`.
+//
+// DataIndex is designed to be passed by value, not reference or pointer. They
+// are also designed to be small and efficient to store in data structures.
+//
+// This templates VectorT on calls instead of using SmallVector explicitly
+// mainly so that the DataType need not be fully declared before declaring
+// DataIndex<DataType>.
+template <typename DataType>
+class DataIndex {
+ public:
+  // Appends an element to the data, and returns the added index.
+  template <typename VectorT>
+  static auto Append(VectorT& data, const DataType& el) -> DataIndex {
+    DataIndex index(data.size());
+    data.push_back(el);
+    return index;
+  }
+  template <typename VectorT>
+  static auto Append(VectorT& data, DataType&& el) -> DataIndex {
+    DataIndex index(data.size());
+    data.push_back(el);
+    return index;
+  }
+
+  // Returns what the index will be if an element is manually added to the data,
+  // after this call.
+  template <typename VectorT>
+  static auto BeforeManualAppend(const VectorT& data) -> DataIndex {
+    return DataIndex(data.size());
+  }
```

Comment:
@zygoloid actually mentioned exactly this kind of approach, along with the `In` methods below, but neither of us were very happy with it.

I think its just a bit too much of an idiomatic break from how vectors typically work.

The design I think he was suggesting was much more around keeping the implicit conversion to an integer type, and just trying to make sure that going from `DataIndex<T>` <-> `DataIndex<U>` is prevented.

Reply:
Okay, but I don't understand why you described that in terms of catching misuse. That is, at best, just reducing keyboard typing.

So does this new approach work for you? If not, I think this is reaching bare minimum levels of utility, so I'm thinking about just closing the PR.